verilog/parser/verilog.y

```diff
@@ -3217,12 +3218,18 @@ block_item_or_statement_or_null_list_opt
   | /* empty */
     { $$ = MakeTaggedNode(N::kBlockItemStatementList); }
   ;
+
+array_range_expression  /* TODO: fill in $$ */
+  : expression
+  | expression ':' expression
+  | expression TK_PO_POS expression
+  | expression TK_PO_NEG expression
+  ;
+
 stream_expression
   : expression
     { $$ = move($1); }
-  /* TODO(fangism):
-  | expression TK_with select_variable_dimension
-   */
+  | expression TK_with_followed_by_bracket array_range_expression ']'
```

Comment:
Related to earlier comment, separate the tokens, but you may give `[` its own flavor.